sound/soc/sdw_utils/soc_sdw_cs_amp.c

```diff
@@ -48,6 +49,46 @@ int asoc_sdw_cs_spk_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai
 }
 EXPORT_SYMBOL_NS(asoc_sdw_cs_spk_rtd_init, SND_SOC_SDW_UTILS);
 
+int asoc_sdw_cs_spk_feedback_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
+{
+	const struct snd_soc_dai_link *dai_link = rtd->dai_link;
+	const struct snd_soc_dai_link_ch_map *ch_map;
+	const struct snd_soc_dai_link_component *codec_dlc;
+	struct snd_soc_dai *codec_dai;
+	u8 ch_slot[8] = {};
+	int amps_per_bus, ch_per_amp, i, ret;
+	unsigned int mask;
+
+	WARN_ON(dai_link->num_cpus > ARRAY_SIZE(ch_slot));
+
+	/*
+	 * CS35L56 has 4 TX channels. When the capture is aggregated the
+	 * same bus slots will be allocated to all the amps on a bus. Only
+	 * one amp on that bus can be transmitting in each slot so divide
+	 * the available 4 slots between all the amps on a bus.
+	 */
+	amps_per_bus = dai_link->num_codecs / dai_link->num_cpus;
+	ch_per_amp = CS_AMP_CHANNELS_PER_AMP / amps_per_bus;
+	WARN_ON(ch_per_amp == 0);
```

Comment:
Any possibility of divide by 0 ? i.e. is the data coming from outside the kernel from topology or ACPI/NHLT ?

Reply:
In theory, it is possible that dai_link->num_codecs < dai_link->num_cpus, and amps_per_bus in that case.
Given that CS_AMP_CHANNELS_PER_AMP is defined as 4, I think ch_per_amp will never be 0.
Maybe change WARN_ON(ch_per_amp == 0); to WARN_ON(amps_per_bus == 0); and move it right after `amps_per_bus = dai_link->num_codecs / dai_link->num_cpus;`?

Reply:
If someone created a system that tried to put 5 amps on each CPU with feedback, there would not be enough TX channels on the amps for each one to have a separate slot. So I think the zero check is legitimate, though unlikely.

But I'm going to handle the possible divide-by-zero that Liam pointed out, and I can combine this with the amps_per_bus check. It's a small overhead to check for out-of-range values.